Ch. 1: Issue 175:  Separate out the PMIx client interface from the CL/host interactions (#192)
pmix/pmix-standard: Chap_Introduction.tex

needed services in distributed and parallel computing systems.
A typical example of such a service is the exchange of information to start a parallel 
application or service in a portable and scalable manner.
As such, \ac{PMIx} also gives distributed system software providers a better understanding of how
programming models and libraries can interface with and use system-level services.
As a specification, \ac{PMIx} also promotes the use and implementation of APIs that allow for
portable access to these varied system software services and the
functionalities they offer.  Although these APIs can be defined and implemented directly by the
system software components providing them, the 
community represented by the Administrative Steering Committee (\ac{ASC})
feels that the development of a shared specification better serves the
community.  Such implementations can interface
with existing system software that are already compliant with \ac{PMIx} or, 
when not already available, directly implement PMIx APIs.
As a result, \ac{PMIx} enables programming languages and libraries to focus on their core
competencies without having to provide their own system-level services.

> \subsection{Background}

Reviewer: ```suggestion
\section{Background}
```

attribute definitions, and communication support for interacting with the local \ac{PMIx} server. 
\ac{PMIx} clients are processes which are started through the \ac{PMIx} infrastructure, 
either by the PMIx implementation directly or through an \ac{SMS} component, and have registered 
as clients.  A \ac{PMIx} client 
is created in such a way that the \ac{PMIx} client library
will be have sufficient information available to
authenticate with the \ac{PMIx} server.  
The \ac{PMIx} server will have sufficient knowledge about the  
process which it created, either directly or through other \ac{SMS}, to authenticate the 
process and provide information the process requests such as its identity and the 
identity of its peers.  

As clients invoke \ac{PMIx} \acp{API}, is possible that some client requests can 
be handled at the client level.  Other requests might require communication with the 
local \ac{PMIx} server, which subsequently might request services from the host \ac{SMS} 
(represented here by an \ac{RM} daemon).  

The interaction between the \ac{PMIx} server and \ac{SMS} might be 
achieved using callback functions registered during server initialization.  

> The host \ac{SMS} could indicate its lack of support for any operation by simply providing a \textit{NULL} for the associated callback function, or could create a function entry that returns \textit{not supported} when called.  These interactions between a \ac{PMIx} implementation 

Reviewer 2: Are these callback functions for the server-side interface?

Author: see above

> and the \ac {SMS} are not defined by the \ac {PMIx} standard.   

Reviewer 2: I thought the `not supported` case is defined by the PMIx standard.

Author: yes, i think that is outdated.  the most recent commits have put this back to show that it is defined

Fig.~\ref{fig:roles} shows how tools can interact with the \ac{PMIx} architecture.
Tools, whether standalone or embedded in job scripts, are an exception to 
the normal client registration process.   A process can register as a tool, provided 
the \ac{PMIx} client library has adequate rendezvous information to connect to the appropriate
\ac{PMIx} server.  This allows processes
which were not created by the PMIx infrastructure to request access to PMIx functionality. 
Processes registered as tools do not have peers.  

\section{Portability of Functionality}
\label{chap:intro:not_supported}

> It is difficult to define a portable \ac{API} that will provice access to the many             

Reviewer: ```suggestion
It is difficult to define a portable \ac{API} that will provide access to the many             
```

Author: done

built against a \ac{PMIx} client library that contains the client-side \acp{API},
attribute definitions, and communication support for interacting with the local \ac{PMIx} server. 
\ac{PMIx} clients are processes which are started through the \ac{PMIx} infrastructure, 
either by the PMIx implementation directly or through a \ac{SMS} component, and have registered 
as clients.  A \ac{PMIx} client 
is created in such a way that the \ac{PMIx} client library
will be have sufficient information available to
authenticate with the \ac{PMIx} server.  
The \ac{PMIx} server will have sufficient knowledge about the  
process which it created, either directly or through other \ac{SMS}, to authenticate the 
process and provide information the process requests such as its identity and the 
identity of its peers.  

As clients invoke \ac{PMIx} \acp{API}, it is possible that some client requests can 
be handled at the client level.  Other requests might require communication with the 
local \ac{PMIx} server, which subsequently might request services from the host \ac{SMS} 
(represented here by a \ac{RM} daemon).  
The interaction between the \ac{PMIx} server and \ac{SMS} are 

> achieved using callback functions registered during server initialization.  

Reviewer 3: An SMS may choose to completely embed the PMIx server; thus not needing the PMIx server API. 

Maybe different wording would work better;
```suggestion
To facilitate the interaction between the \ac{PMIx} server and the \ac{SMS}, the \ac{PMIx} server API provides callback functions to be registered during server initialization.
```

required using a flag within the \refstruct{pmix_info_t} (See \ref{chap:struct:info}). 
If the required attribute is 
not available on the system or the desired value for the attribute is not available, the call will 
return the error code for \textit{not supported}.

For example, the \refattr{PMIX_TIMEOUT} attribute can be used to specify the time (in seconds) before the requested operation should time out. The intent of this attribute is to allow the client to avoid ``hanging'' in a request that takes longer than the client wishes to wait, or may never return (e.g., a \refapi{PMIx_Fence} that a blocked participant never enters).

The application can query the attribute support for \refapi{PMIx_Fence} 
and search whether \refattr{PMIX_TIMEOUT} is listed as a supported attribute.  The application can also   
set the required flag in the \refstruct{pmix_info_t} for that attribute when making the \refapi{PMIx_Fence} 
call. This will return an error if this attribute is not supported. If the required flag is not set, 
the library and \ac{SMS} host are allowed to treat the attribute as optional, ignoring it if support 
is not available.

> It is therefore critical that users and application implementers:

Reviewer 3: Should start with a definition of what is the PRI

is created in such a way that the \ac{PMIx} client library
will be have sufficient information available to
authenticate with the \ac{PMIx} server.  
The \ac{PMIx} server will have sufficient knowledge about the  
process which it created, either directly or through other \ac{SMS}, to authenticate the 
process and provide information the process requests such as its identity and the 
identity of its peers.  

As clients invoke \ac{PMIx} \acp{API}, it is possible that some client requests can 
be handled at the client level.  Other requests might require communication with the 
local \ac{PMIx} server, subsequently might request services from the host \ac{SMS} 
(represented here by a \ac{RM} daemon).  
The interaction between the \ac{PMIx} server and \ac{SMS} are 
achieved using callback functions registered during server initialization.  
The host \ac{SMS} can indicate its lack of support for any operation by simply providing a \textit{NULL} for the associated callback function, or can create a function entry that returns \textit{not supported} when called.

Recognizing the burden this places on SMS vendors, the PMIx community has included interfaces by

> which the host can request support from local SMS elements. Once the \ac{SMS} has transferred the request to

Reviewer: ```suggestion
which the host \ac{SMS} (containing the local PMIx service instance) can request support from local \ac{SMS} elements via the PMIx API. Once the \ac{SMS} has transferred the request to
```

an appropriate location, a \ac{PMIx} server interface can be used to pass the request 
between \ac{SMS} subsystems.
For example, a request for network traffic statistics can utilize the
PMIx networking abstractions to retrieve the information from the Fabric Manager. 
This reduces the portability and
interoperability issues between the individual subsystems by transferring the burden of defining the
interoperable interfaces from the \ac{SMS} subsystems to the \ac{PMIx} community, which continues
to work with those providers to develop the necessary support.

Fig.~\ref{fig:roles} shows how tools can interact with the \ac{PMIx} architecture.
Tools, whether standalone or embedded in job scripts, are an exception to 
the normal client registration process.   A process can register as a tool, provided 
the \ac{PMIx} client library has adequate rendezvous information to connect to the appropriate

> \ac{PMIx} server.  This allows processes

Reviewer: ```suggestion
\ac{PMIx} server (either hosted on the local machine or on a remote machine).  This allows processes
```

which were not created by the PMIx infrastructure to request access to PMIx functionality. 

> Processes registered as tools do not have peers.  

Reviewer: Per discussion on teleconf - strike this last sentence. It _may_ be possible that tools have peers (see `PMIX_TOOL_NSPACE` / `PMIX_TOOL_RANK` in the `PMIx_tool_init` call)

could be so burdensome as to discourage \ac{PMIx} implementations.

To help address this issue, the \ac{PMIx} \acp{API} are designed to allow resource managers 
and other system management stack components to decide on support of a 
particular function and allow client applications to query and adjust to the level of support available.  \ac{PMIx} clients should be written to account for the possibility that a \ac{PMIx} \ac{API} might return an error code indicating that the call is not supported.
The \ac{PMIx} community continues to look at ways to assist \ac{SMS} implementers in their decisions 
on what functionality to support by highlighting functions and attributes that are 
critical to basic application execution (e.g., \refapi{PMIx_Get}) for certain classes of applications.

\subsection{Optional Nature of Attributes}

An area where differences between support on different systems can be challenging is regarding the attributes that provide information to the client process and/or control the behavior of a \ac{PMIx} \ac{API}.  Most
\ac{PMIx} \ac{API} calls can accept additional information or attributes specified in the form of 

> key/value pairs. These attributes provide information to the \ac{PMIx} implementation which influence the 

Reviewer: ```suggestion
key/value pairs. These attributes provide information to the \ac{PMIx} implementation that influence the 
```

process and provide information the process requests such as its identity and the 
identity of its peers.  

As clients invoke \ac{PMIx} \acp{API}, it is possible that some client requests can 
be handled at the client level.  Other requests might require communication with the 

> local \ac{PMIx} server, subsequently might request services from the host \ac{SMS} 

Reviewer: ```suggestion
local \ac{PMIx} server, which subsequently might request services from the host \ac{SMS} 
```

interoperable interfaces from the SMS subsystems to the PMIx community, which continues
to work with those providers to develop the necessary support.

\ac{PMIx} clients are processes which are started through the \ac{PMIx} infrastructure, 
either by the PMIx implementation directly or through an \ac{SMS} component, and have registered 
as clients.  A \ac{PMIx} client 
is created in such a way that the \ac{PMIx} client library
will be have sufficient information available to
authenticate with the \ac{PMIx} server.  
The \ac{PMIx} server will have sufficient knowledge about the  
process which it created, either directly or through other \ac{SMS}, to authenticate the 
process and provide information the process requests such as its identity and the 
identity of its peers.  

Tools, whether standalone or embedded in job scripts, are an exception to 
the normal client registration process.   A process can register as a tool, provided 
the \ac{PMIx} client library has adequate rendezvous information to connect to the appropriate

> \ac{PMIc} server.  This allows processes

Reviewer: ```suggestion
\ac{PMIx} server.  This allows processes
```

Author: done
